script.module.codequick/lib/codequick/route.py

```diff
@@ -0,0 +1,146 @@
+import logging
+import inspect
+import re
+
+# Kodi imports
+import xbmcplugin
+
+# Package imports
+from codequick.script import Script
+from codequick.support import logger_id, auto_sort
+from codequick.utils import ensure_native_str
+
+__all__ = ["Route"]
+
+# Logger specific to this module
+logger = logging.getLogger("%s.route" % logger_id)
+
+# Localized string Constants
+SELECT_PLAYBACK_ITEM = 25006
+NO_DATA = 33077
+
+
+class Route(Script):
+    """
+    This class is used to create Route callbacks. Route callbacks, are callbacks that
+    return listitems witch show up as folders in kodi.
+
+    Route inherits all methods and attributes from :class:`script.Script<codequick.script.Script>`.
+
+    :example:
+        >>> from codequick import Route, Listitem
+        >>>
+        >>> @Route.register
+        >>> def root(_):
+        >>>     yield Listitem.from_dict("Extra videos", subfolder)
+        >>>     yield Listitem.from_dict("Play video",
+        >>>           "http://www.example.com/video1.mkv")
+        >>>
+        >>> @Route.register
+        >>> def subfolder(_):
+        >>>     yield Listitem.from_dict("Play extra video",
+        >>>           "http://www.example.com/video2.mkv")
+    """
+
+    # Change listitem type to 'folder'
+    is_folder = True
+
+    def __init__(self):
+        super(Route, self).__init__()
+        self.update_listing = self.params.get(u"_updatelisting_", False)
+        self._manual_sort = set()
+        self.content_type = None
+        self.autosort = True
+
+    def _execute_route(self, callback):
+        """Execute the callback function and process the results."""
+
+        # Fetch all listitems from callback function
+        listitems = super(Route, self)._execute_route(callback)
+
+        # Process listitems and close
+        success = self.__add_listitems(listitems)
+        self.__end_directory(success)
+
+    def __add_listitems(self, raw_listitems):
+        """Handle the processing of the listitems."""
+        # Convert listitem to list incase we have a generator
+        if inspect.isgenerator(raw_listitems):
+            raw_listitems = list(raw_listitems)
+
+        # If raw_listitems is False then, that was deliberate, so return False
+        if raw_listitems is False:
+            return False
+
+        # Check if raw_listitems is None or an empty list
+        elif not raw_listitems:
+            raise RuntimeError("No items found")
+
+        # Create a new list containing tuples, consisting of path, listitem, isfolder.
+        listitems = []
+        folder_counter = 0.0
+        for listitem in raw_listitems:
+            if listitem:
+                # noinspection PyProtectedMember
+                listitem_tuple = listitem._close()
+                listitems.append(listitem_tuple)
+                if listitem_tuple[2]:
+                    folder_counter += 1
+
+        # Guess if this directory listing is primarily a folder or video listing.
```

Comment:
Not sure if this is the right way to do this. You should be setting the infolabel "mediatype" for each listitem to the correct value. I'm not sure if you're doing this because of mixed directory or you have something else in mind here.

Reply:
listitem_tuple contains (path, xbmcgui.listitem, isfolder).
This checks if it is listing folders or video items, so to set xbmcplugin.setContent differently.
"files" when listing folders and "videos" when listing media items.

Reply:
I can't see where you are setting the "mediatype" label which is used by newer skins.

Reply:
Sorry. Wasn't sure what you were on about. To be honest I forgot about "mediatype". It would be set by the add-ons that use codequick. If needed I could set "mediatype" to default to "video", if it was not set by the add-on. I will change that bit of code to take the infolabel "mediatype" into consideration.